### DISCUSSION TITLE
EChart toolbox event

### DISCUSSION THREAD

**Discussion by CSharpDummy** [zauberzeug/nicegui#2679] [Ideas / Feature Requests]
### Question

I need to capture the toolbox select event in EChart. Idea is to have the user select an area in a time series and the run some analytics on it. The on click event works well to select points on the chart, but I'm having issues wiring up the toolbox event. I tried:
```
    "toolbox": {
        "feature": {
            "dataZoom": {},
            "brush": {
                "type": ["rect", "polygon", "clear"],
                "onclick:": "function () {emitEvent('tb_selected');}",
            },
        }
    },
```

but that didn't work. Any ideas?

**Comment by falkoschindler** [comment]
@CSharpDummy Can you, please, share a minimal reproducible example on how to add the toolbox to the chart? I tried copying your code into the demo from the website, but I don't see the "brush" at all. Thanks!

**Reply by CSharpDummy** [reply]
Here is the minimum nicegui code:

Thx for your help!
```py
#!/usr/bin/env python3
from nicegui import ui

data1 = [
    [161.2, 51.6],
    [167.5, 59.0],
    [159.5, 49.2],
]

data2 = [
    [174.0, 65.6],
    [175.3, 71.8],
    [193.5, 80.7],
]

OPTION = {
    "toolbox": {
        "feature": {
            "dataZoom": {},
            "brush": {
                "type": ["rect", "polygon", "clear"],
                "onclick:": "function () {emitEvent('tb_selected');}",
            },
        }
    },
    "brush": {},
    "legend": {"data": ["Female", "Male"], "left": "center", "bottom": 10},
    "xAxis": [
        {
            "type": "value",
            "scale": True,
            "axisLabel": {"formatter": "{value} cm"},
            "splitLine": {"show": False},
        }
    ],
    "yAxis": [
        {
            "type": "value",
            "scale": True,
            "axisLabel": {"formatter": "{value} kg"},
            "splitLine": {"show": False},
        }
    ],
    "series": [
        {
            "name": "Female",
            "type": "scatter",
            "emphasis": {"focus": "series"},
            "data": data1,
            "markArea": {
                "silent": True,
                "itemStyle": {
                    "color": "transparent",
                    "borderWidth": 1,
                    "borderType": "dashed",
                },
                "data": [
                    [
                        {"name": "Female Data Range", "xAxis": "min", "yAxis": "min"},
                        {"xAxis": "max", "yAxis": "max"},
                    ]
                ],
            },
            "markPoint": {
                "data": [{"type": "max", "name": "Max"}, {"type": "min", "name": "Min"}]
            },
            "markLine": {
                "lineStyle": {"type": "solid"},
                "data": [{"type": "average", "name": "AVG"}, {"xAxis": 160}],
            },
        },
        {
            "name": "Male",
            "type": "scatter",
            "emphasis": {"focus": "series"},
            "data": data2,
            "markArea": {
                "silent": True,
                "itemStyle": {
                    "color": "transparent",
                    "borderWidth": 1,
                    "borderType": "dashed",
                },
                "data": [
                    [
                        {"name": "Male Data Range", "xAxis": "min", "yAxis": "min"},
                        {"xAxis": "max", "yAxis": "max"},
                    ]
                ],
            },
            "markPoint": {
                "data": [{"type": "max", "name": "Max"}, {"type": "min", "name": "Min"}]
            },
            "markLine": {
                "lineStyle": {"type": "solid"},
                "data": [{"type": "average", "name": "Average"}, {"xAxis": 170}],
            },
        },
    ],
}
echart = ui.echart(OPTION, on_point_click=lambda e: ui.notify(f"Clicked on {e}"))

ui.run()
```

**Reply by falkoschindler** [reply]
Thanks, @CSharpDummy! Changing `"onclick:"` to `":onclick"` _should_ do the trick, but it isn't working. Using the JavaScript console I could verify that the function is correctly converted to a function object. But it isn't called, which I tested using a `console.log` instead of `emitEvent`.
I'm still not sure if we use the "toolbox" API correctly. Maybe you can try to create a plain HTML example with a working click handler for comparison?

**Reply by CSharpDummy** [reply]
I played around with it and what I really need get the coordinates of the selections. There is "brush" event, but this is not supported yet:

```js
myChart.on('brush', function(params) {
  console.log("brush");
  console.log(params)
});
```

Let me know how can I help to get it into the repo?

**Reply by CSharpDummy** [reply]
You could use plotly that has a lot more events exposed. There is also one to return the current selection.

**Reply by falkoschindler** [reply]
Ah ok, I was mislead by the `onclick:` attribute in the options dictionary. But registering a "brush" event like we do for "click" here
https://github.com/zauberzeug/nicegui/blob/1911455622d537a22b16778f2b7d244d526f125c/nicegui/elements/echart.js#L7
would probably solve your problem.

While at it, we could add support for even more events. Do you know where to find a list of all supported events? [Here](https://api.highcharts.com/highcharts/chart.events) they don't mention the "brush" event, even though it works.

**Reply by falkoschindler** [reply]
Oh! Here it is: https://echarts.apache.org/en/api.html#events. I confused it with Highcharts.

**Reply by falkoschindler** [reply]
I just created PR https://github.com/zauberzeug/nicegui/pull/2716 to add support for all these chart events.

**Reply by CSharpDummy** [reply]
@falkoschindler this works great! Makes the plot much more interactive.

**Comment by falkoschindler** [comment]
This feature request has been released in version 1.4.19. ✔️
